Demo version

KatyaKos opened the pull request:
Нет базы данных, нет геолокации, нет загрузки картинок из галереи, нет работы с фильмами и книгами. Сервера тоже нет.

adkozlov commented:
1. Жуткая работа с репозиторием. Бинарные файлы, генерируемые файлы... Очень неудобно выделять реальные изменения.
2. Неинформативные commit-messages. Пустые или не относящиеся к изменяемым файлам.
3. Ветку нужно переделать и сделать новый реквест.
4. Выкладывание картинок в таком виде является нарушением авторских прав. Сделать в соответствии с юридическими нормами или избавиться от них.
5. В коде много копипасты. Например, абсолютно одинаковый код инициализации кнопок написан раз 6.
6. Достаточно длинные методы. Выделить содержательные подметоды, вынести в качестве утилитарных (скорее всего статических). Очень помогает встроенный рефакторинг _Extract method_.
7. Класс `LoginActivity` практически полностью дублирует функциональность класса `SignupActivity`.
8. Такого [сайта](http://spbau.me) не существует. Пакеты должны быть переименованы.

Более

adkozlov commented:
И еще одно уточнение.  `README.md` должен лежать _внутри_ проекта. Ну и дополнительный уровень вложение проекта не нужен.

KatyaKos committed:
project create

KatyaKos committed:
Login Activity create.
Contains fields for email and password, login button and link to signup.

KatyaKos committed:
Signup Activity create.
Contains fields for email, login and password, signup button and link to Login Activity.

KatyaKos committed:
Main Activity create.
Contains "move between Activities" buttons.

KatyaKos committed:
Profile Activity create.
Contains information about user and settings button.

KatyaKos committed:
ProfileSettings Activity create.
Changes information about user. Contains save button and information fields.

KatyaKos committed:
Item for Friends, Movies and Books listviews.
Contains image, two textfields and button.

KatyaKos committed:
Friends Activity create.
Contains list of friends, button "move to Add Friends".

KatyaKos committed:
AddFriends Activity create.
Contains list of all users, search field, button add for each user who is not your friend, button delete for friends.

KatyaKos committed:
Added structure with information about all users.

KatyaKos committed:
Added structure with information about one user.

KatyaKos committed:
Diary Activity create.
Contains user's photo, name and nickname. Contains list of notes, "tag search" and "add new note" button.

KatyaKos committed:
Added note listview item.
Contains fields for note's name, content and tags (using for filter).

KatyaKos committed:
AddNote Activity create.
Contains fields for note's name, content and tags (using for filter). Contains add button.

KatyaKos committed:
Map Activity create.
Contains list of groups (trips) with places inside, buttons add new group and add new place, delete buttons.

KatyaKos committed:
Added items for Map expandable listview.
Groups' item contains name of group and buttons add, delete.
Places' item contains name of place and delete button.

KatyaKos committed:
Movies Activity create.
Contains list of movies, sort spinner, filter button, add and delete movies buttons.

KatyaKos committed:
Books Activity create.
Contains list of books, sort spinner, filter button, add and delete books buttons.

KatyaKos committed:
Added item for sort spinner.
Used in Movies and Books sort.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+            "foo@example.com:hello", "bar@example.com:world"
+    };
+
+    private static final int REQUEST_SIGNUP = 0;
```
В чем суть данной константы? Если она планируется на будущее -- это нужно пометить.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_login);
+
+        emailText = (EditText) findViewById(R.id.login_email);
```
Можно использовать библиотеку ButterKnife для инъекции объектов.
Это совершенно необязательно, но было бы хорошо.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+
+    private static final int REQUEST_SIGNUP = 0;
+
+    // UI references.
```
Непонятный и ненужный комментарий.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+        String email = emailText.getText().toString();
+        String password = passwordText.getText().toString();
+
+        if (email.isEmpty() || !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
```
EMAIL_ADDRESS можно статически заимпортировать. Тогда не будет такой длинной строчки.

adkozlov replied:
Проблема этого метода заключается в том, что работа с логикой и UI должны быть разделены.

Для логина и пароля разные предикаты. Но действие над EditText объектами абсолютно одинаковые.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+            emailText.requestFocus();
+            valid = false;
+        } else {
+            emailText.setError(null);
```
Зачем это нужно?

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+        }
+
+        if (password.isEmpty() || password.length() < 4 || password.length() > 10) {
+            passwordText.setError("between 4 and 10 alphanumeric characters");
```
Еще одной проблемой данного метода является наличие сайд-эффектов.
Проверка корректности логина/пароля не подразумевает изменение какого-либо состояния (ну если только не пишется система, где запоминаются любые попытки проникнуть).
А вот уже работа с UI -- это сайд-эффект.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+    private boolean validate() {
+        boolean valid = true;
+
+        String email = emailText.getText().toString();
```
TextView::getText().toString() в **одном** классе написано 4 раза!

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+            emailText.setError(null);
+        }
+
+        if (password.isEmpty() || password.length() < 4 || password.length() > 10) {
```
Мне кажется, 0 < 4.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    private boolean validate() {
+        boolean valid = true;
```
Булевские флаги принято называть с префиксом _is_.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+
+    private boolean checkEmailPassword(String email, String password) {
+        for (String credential : DUMMY_CREDENTIALS) {
+            String[] pieces = credential.split(":");
```
Зачем так сложно? Можно было хранить массив массивов по 2 элемента.
Я уж не говорю о том, что есть класс Pair.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+
+    public void login() {
```
Почему метод публичный?
login -- плохое название для того, что внутри происходит.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+
+
+    public void login() {
+
```
Пустая строчка.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+        // TODO: Implement your own authentication logic here.
+
+        if (checkEmailPassword(email, password)) {
+            progressDialog.cancel();
```
Получается, что cancel() вызывается в любом случае.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+
+        loginButton.setEnabled(false);
+
+        final ProgressDialog progressDialog = new ProgressDialog(LoginActivity.this, R.style.AppTheme);
```
В этом методе понамешано всего подряд. В комментарии к методу validate() можно прочитать про разделение логики и UI.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+            startActivity(intent);
+        } else {
+            progressDialog.cancel();
+            passwordText.setError(getString(R.string.error_incorrect_password));
```
Почему же тогда все остальные сообщения не описаны константами?

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+            if (resultCode == RESULT_OK) {
+
+                // TODO: Implement successful signup logic here
+                this.finish();
```
Где-то я такой же код уже видел... 17 строк назад?

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/LoginActivity.java:
```diff
@@ -0,0 +1,167 @@
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == REQUEST_SIGNUP) {
```
В целом объединение двух логических выражений в одно выглядит гораздо лучше, чем еще один уровень смещения.

adkozlov commented on app/src/main/AndroidManifest.xml:
```diff
@@ -1,13 +1,29 @@
+                <category android:name="android.intent.category.LAUNCHER" />
+            </intent-filter>
+        </activity>
+        <activity
```
Тут проблема. Про SignupActivity в данном коммите еще ничего не известно.

adkozlov commented on app/proguard-rules.pro:
```diff
@@ -0,0 +1,17 @@
+# Add project specific ProGuard rules here.
+# By default, the flags in this file are appended to flags specified
+# in C:\Users\KatyaKos\AppData\Local\Android\Sdk/tools/proguard/proguard-android.txt
```
В репозитории не должно быть ненужных файлов.
А файлы тем более не должны содержать путей на компьютере разработчика.

adkozlov commented on gradlew:
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
```
Не нужно коммитить исполняемые файлы gradle. Он есть у меня на компьютере.
Уж тем более jar-файлы! Жаль к нему нельзя комментарии писать.

adkozlov commented on .gitignore:
```diff
@@ -34,7 +34,7 @@ captures/
 # Intellij
 *.iml
-.idea/workspace.xml
+.idea/
```
Commit-message плохой. _Created_ хотя бы.

adkozlov commented on app/src/main/AndroidManifest.xml:
```diff
@@ -1,13 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
```
То же самое. Commit-message обычно пишется в Past Simple.

adkozlov commented on app/src/main/AndroidManifest.xml:
```diff
@@ -21,6 +21,7 @@
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+
```
Проблема из предыдущего коммита.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/SignupActivity.java:
```diff
@@ -0,0 +1,131 @@
+ * Created by KatyaKos on 14.11.2016.
+ */
+
+public class SignupActivity extends Activity {
```
Практически весь данный класс является копипастой LoginActivity.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/SignupActivity.java:
```diff
@@ -0,0 +1,131 @@
+        progressDialog.setMessage("Creating Account...");
+        progressDialog.show();
+
+        // TODO: Implement your own signup logic here.
```
Выглядит, как будто бы я должен написать эту логику здесь.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/SignupActivity.java:
```diff
@@ -0,0 +1,131 @@
+        return true;
+    }
+
+    private boolean validate() {
```
Все те же самые проблемы.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/MainActivity.java:
```diff
@@ -0,0 +1,141 @@
+
+public class MainActivity extends Activity {
+
+    private static final int REQUEST_CHANGE = 0;
```
Снова непонятная константа.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/MainActivity.java:
```diff
@@ -0,0 +1,141 @@
+        buttonsListener();
+    }
+
+    private void initialization() {
```
Не могу сказать, что выделение этих строчек в отдельный метод добавляет читаемости.
Всю инициализацию можно в onCreate() писать.

Отделить просто блоки друг от друга пустыми строчками.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/MainActivity.java:
```diff
@@ -0,0 +1,141 @@
+        TextView userNickname = (TextView) findViewById(R.id.main_user_nickname);
+        TextView userName = (TextView) findViewById(R.id.main_user_name);
+
+        int id = getResources().getIdentifier(UserInformation.getUserPhoto(), "drawable", getPackageName());
```
id в данном случае не самый лучший идентификатор. Можно было добавить что-то относящееся к картинке.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/MainActivity.java:
```diff
@@ -0,0 +1,141 @@
+
+    private void initialization() {
+        intentMain = new Intent(getApplicationContext(), MainActivity.class);
+        intentProfile = new Intent(getApplicationContext(), ProfileActivity.class);
```
И вот тут начинается самое интересное.
В рамках данного коммита я знаю только про MainActivity, SignupActivity и LoginActivity.
Откуда берутся все остальные классы, почему они используются?

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/MainActivity.java:
```diff
@@ -0,0 +1,141 @@
+        userName.setText(UserInformation.getUserName());
+    }
+
+    private void buttonsListener() {
```
Listener по всей видимости один на всех. И все на одного.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/MainActivity.java:
```diff
@@ -0,0 +1,141 @@
+            }
+        });
+
+        profileButton.setOnClickListener(new View.OnClickListener() {
```
Продолжение проблемы. Я все еще понятия не имею ни про какие activity кроме трех перечисленных.

И вот если бы все было правильно с точки зрения истории, то было бы понятно следующее.
Допустим, в какой-то момент появляется ProfileActivity. Ему будет соответствовать indent и кнопка.
Проинициализировали indent и тут же добавили listener соответствующей кнопке.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/MainActivity.java:
```diff
@@ -0,0 +1,141 @@
+            }
+        });
+
+        friendsButton.setOnClickListener(new View.OnClickListener() {
```
Как только появляется еще одно activity становится понятно, что этот алгоритм выполняется для **всех** activity.
И уже на втором можно чуть-чуть обобщить логику и избавиться от всего этого страшного boiler-plate кода.

Инициализация всех элементов, соответствовавших новой activity происходит в одном методе. И onCreate() может выглядеть как-то так:

`
registerButtonFor(ProfileActivity.class); registerButtonFor(FriendsActivity.class);
`
И так далее.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileActivity.java:
```diff
@@ -0,0 +1,99 @@
+        menuButton = (Button) findViewById(R.id.profile_button_menu);
+    }
+
+    private void fieldsInitialization() {
```
Тут все примерно то же самое.
Можно долго спорить о том, в каком порядке расположить строчки в onCreate().
Лично мне кажется логичным, когда объекты инициализируются по очереди.

Взяли headerImage, добавили картинку, закончили с ним, перешли к profileImage.
И тогда можно на методы разбить с другой группировкой.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileActivity.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private void fieldsInitialization() {
+        int idHeader = getResources().getIdentifier(UserInformation.getUserHeader(), "drawable", getPackageName());
```
Очень длинная строчка.

adkozlov replied:
Все то же самое: что такое UserInformation? Откуда оно? Что у него за статические методы?

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileActivity.java:
```diff
@@ -0,0 +1,99 @@
+    private void fieldsInitialization() {
+        int idHeader = getResources().getIdentifier(UserInformation.getUserHeader(), "drawable", getPackageName());
+        headerImage.setImageResource(idHeader);
+        int idPhoto = getResources().getIdentifier(UserInformation.getUserPhoto(), "drawable", getPackageName());
```
Очень длинная _такая же_ строчка.
Отлично параметризуется.

adkozlov replied:
Лучше бы хотя бы photoId.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+                UserInformation.setUserBio(bioText.getText().toString());
+                UserInformation.setUserName(nameText.getText().toString());
+                UserInformation.setUserBirth(birthText.getText().toString());
+                UserInformation.setUserAbout(aboutText.getText().toString());
```
В пятый раз одно и то же должно быть писать лень.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+            nicknameText.setError("nickname is too short");
+            return false;
+        } else if (nickname.contains("@")) {
+            nicknameText.setError("don't use \"@\" symbols");
```
Страдательный залог, глаголы avoid или forbid.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+        } else {
+            nickname = "@" + nickname;
+            if (!nickname.equals(UserInformation.getUserNickname()) && AllUsersInformation.usersContain(nickname)) {
+                nicknameText.setError("nickname is busy");
```
Google translate? 😄

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+ * Created by KatyaKos on 27.11.2016.
+ */
+
+public class ProfileSettingsActivity extends Activity {
```
Класс брат-близнец ProfileActivity.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+    private EditText birthText;
+    private EditText aboutText;
+
+    private ImageView headerImage;
```
Не использующиеся поля.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+        aboutText.setText(UserInformation.getUserAbout());
+    }
+
+    private void fieldsInitialization() {
```
Не использующийся метод.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+        int idPhoto = getResources().getIdentifier(UserInformation.getUserPhoto(), "drawable", getPackageName());
+        profileImage.setImageResource(idPhoto);*/
+
+        nicknameText.setText(UserInformation.getUserNickname());
```
Логика метода должна быть следующей. На вход принимается объекта типа UserInformation (Проще назвать его User).
И по этому информации из этого объекта заполняются EditText поля.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+                    Toast.makeText(getBaseContext(), "Change nickname", Toast.LENGTH_LONG).show();
+                    return;
+                }
+                UserInformation.setUserNickname(nicknameText.getText().toString());
```
Здесь логика должна быть следующая. Есть логин валиден, создаем UserInformation по текстовым полям. И отправляем его в некоторый менеджер.

AllUsersInformation -- это и есть реализация идеи менеджера как некоторого общего хранилища. После того, как добавлен этот уровень абстракции, реализация легко меняется. Это может быть просто синглтон с ассоциативной структурой, а может быть и база данных.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/ProfileSettingsActivity.java:
```diff
@@ -0,0 +1,121 @@
+            return false;
+        } else {
+            nickname = "@" + nickname;
+            if (!nickname.equals(UserInformation.getUserNickname()) && AllUsersInformation.usersContain(nickname)) {
```
Обычно такие менеджеры (реестры) предоставляют следующий интерфейс:
`void addUser(User user); User findUserByName(String name); void removeUser(какой-то способ указания на удаляемый объект)`.

В данном случае нам нужен метод find. Он может быть просто nullable, может возвращать Option, это не важно.
Важно, что в этой строчке должно быть что-то типа: `UserRegistry.findUserByName(nickName) != null`.
Что соответствует тому факту, что пользователь под данным ником уже существует.

adkozlov reviewed (changes requested):
Появилась некоторая сущность, а зависящие от нее места не появились.

KatyaKos committed:
Changed gitignore.

Added proguard and graddle files.

KatyaKos committed:
Deleted extra files

KatyaKos committed:
Deleted extra files

KatyaKos committed:
Fixed mistakes

Login function is divided into loginUI and loginLogic

KatyaKos committed:
Login fixed mistakes

loginUI() fixed

KatyaKos committed:
Fixed Main mistakes

registerButton*() functions added

adkozlov commented on app/src/main/res/layout/item_listview.xml:
```diff
@@ -0,0 +1,53 @@
+    android:layout_marginBottom="15dp"
+    android:background="#fff">
+
+    <ImageView
```
Сущности, которые нигде не используются (в рамках коммита).

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/FriendsActivity.java:
```diff
@@ -0,0 +1,140 @@
+
+    private static final int REQUEST_ADD = 0;
+
+    Button menuButton;
```
Публичные поля.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/FriendsActivity.java:
```diff
@@ -0,0 +1,140 @@
+
+public class FriendsActivity extends Activity {
+
+    private static final int REQUEST_ADD = 0;
```
Снова приватная константа.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/FriendsActivity.java:
```diff
@@ -0,0 +1,140 @@
+        friendsList = (LinearLayout) findViewById(R.id.friends_list);
+    }
+
+    private void friendsListCreate(CharSequence sort) {
```
Почему используется CharSequence, который далее приводится к String?
Зачем нужен этот параметр, если он одинаковый в двух вызовах из трех?

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/FriendsActivity.java:
```diff
@@ -0,0 +1,140 @@
+
+        for (String friend : userFriends) {
+            String[] pieces = friend.split("::");
+            if (!(pieces[0].toLowerCase().contains(sort.toString().toLowerCase()) ||
```
Проверка на то, что какая-то строка содержит пустую подстроку?
Опять копипаста вместо выделенных методов.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/FriendsActivity.java:
```diff
@@ -0,0 +1,140 @@
+        LayoutInflater layoutInflater = getLayoutInflater();
+
+        for (String friend : userFriends) {
+            String[] pieces = friend.split("::");
```
Опять проблемы из-за странной идеи хранения данных.
Эта логика должна быть написана ровно один раз.

adkozlov commented on app/src/main/java/me/spbau/katyakos/android/museumofme/FriendsActivity.java:
```diff
@@ -0,0 +1,140 @@
+            photo.setImageResource(id);
+
+            TextView nickname = (TextView) item.findViewById(R.id.listview_text1);
+            nickname.setText(pieces[0]);
```
Опять смешение модели и UI.

adkozlov commented:
Пункты, которые нужно поправить к следующей попытке сдачи:
1. вся так называемая бизнес-логика должна быть продумана и реализована в виде содержательных сущностей и соответствующих им классов;
2. вся содержательная логика должна быть реализована ровно один раз, а не в каждом месте заново;
3. UI должен быть полностью отделен от логики (в идеале должно принять форму одного из MV-паттернов, скорее всего MVP);
4. реализации должны быть легко заменяемы (например, переход от хранения всех данных в текстовом файле к хранению в базе не должен требовать замены половины классов);
5. низкоуровневый код с полным отсутствием абстракций (aka "олимпиадный код") приниматься не будет.

И еще один совет: приложение должно расширяться методом добавления функциональности, а не беспорядочным добавлением недоделанных блоков.

Для того, чтобы увеличить продуктивность сдач и вообще вероятность получения зачета, предлагаю ввести систему mileston'ов.
В качестве первого предлагается реализовать приложение, в котором можно зарегистрироваться и в которое можно войти. Данные пользователей должны храниться в едином реестре (текстовый файл/база данных/mock-объект).
